Add a way to know if a CachedImage is showing a complete frame

smfr opened the pull request:
#### 0fee98d9aebfaae332f980001c3325824379474f
<pre>
Add a way to know if a CachedImage is showing a complete frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=299647">https://bugs.webkit.org/show_bug.cgi?id=299647</a>
<a href="https://rdar.apple.com/161453244">rdar://161453244</a>

Reviewed by Said Abou-Hallawa.

LargestContentfulPaint needs to know if an image is showing a fully-decoded frame.
For an animated image, this can mean that the entire image is still downloading
(so we can&apos;t check the status()), but if the current frame is complete, we can count it for LCP.

For a single-frame image, &quot;complete&quot; is equivalent to the image download being complete.

Fix by exposing ImageSource&apos;s `currentImageFrame().isComplete()` up through to CachedImage,
for bitmap images. For other image types, choose something reasonable.

This will be tested by image-related largest-contentful-paint tests.

* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::currentFrameIsComplete):
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/GeneratedImage.h:
* Source/WebCore/platform/graphics/Image.h:
* Source/WebCore/platform/graphics/cg/PDFDocumentImage.h:
* Source/WebCore/rendering/RenderImageResource.cpp:
(WebCore::RenderImageResource::currentFrameIsComplete const):
* Source/WebCore/rendering/RenderImageResource.h:
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGImageForContainer.h:

Canonical link: <a href="https://commits.webkit.org/300641@main">https://commits.webkit.org/300641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14bb2ba31fd8cdd36b89bee3043a75c7860be159

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123329 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130045 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/28d476a6-f312-4874-b807-755d39af1bd2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43767 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51638 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93748 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0ee12826-ff23-4c9d-a06f-a32f1f9c8841) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126282 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34874 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110345 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74378 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d1fe118b-c3a7-42a1-9d4c-df024ac70986) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33840 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28504 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73558 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104588 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28730 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132761 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50279 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38261 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102242 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50655 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106569 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102096 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25955 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47445 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25667 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47065 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50134 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55895 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49605 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52955 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51283 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

shallawa reviewed (approved):
The change does what the commit message says. But I am not sure exactly what you are trying to achieve eventually.

For the animated image, we do not start animating it until the image is fully loaded. This is because the number of frames can't be final until all the data is cached. But this does not prevent us from decoding the first frame which may be complete even though loading the image is still in progress.

To know whether the image encoded data is fully loaded (either static or animating) I think you can check 
```
    if (cachedImage.status() == CachedResource::Status::Cached)
```

smfr committed:
Add a way to know if a CachedImage is showing a complete frame
https://bugs.webkit.org/show_bug.cgi?id=299647
rdar://161453244

Reviewed by Said Abou-Hallawa.

LargestContentfulPaint needs to know if an image is showing a fully-decoded frame.
For an animated image, this can mean that the entire image is still downloading
(so we can't check the status()), but if the current frame is complete, we can count it for LCP.

For a single-frame image, "complete" is equivalent to the image download being complete.

Fix by exposing ImageSource's `currentImageFrame().isComplete()` up through to CachedImage,
for bitmap images. For other image types, choose something reasonable.

This will be tested by image-related largest-contentful-paint tests.

* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::currentFrameIsComplete):
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/GeneratedImage.h:
* Source/WebCore/platform/graphics/Image.h:
* Source/WebCore/platform/graphics/cg/PDFDocumentImage.h:
* Source/WebCore/rendering/RenderImageResource.cpp:
(WebCore::RenderImageResource::currentFrameIsComplete const):
* Source/WebCore/rendering/RenderImageResource.h:
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGImageForContainer.h:

Canonical link: https://commits.webkit.org/300641@main
